espnet3/runner/base_runner.py

```diff
@@ -0,0 +1,392 @@
+            for res in parallel_for(
+                self.__class__.forward, indices, setup_fn=setup_fn, client=client
+            ):
+                out.append(res)
+        return out
+
+    async def _run_async(self, indices: Sequence[int]) -> List[Any] | None:
+        """Submit shards to a Dask cluster and gather or write results.
+
+        Workflow:
+            1) Emit per-shard JSON specs to ``async_specs_dir``.
+            2) Submit worker entrypoints that reconstruct runner/provider and run.
+            3) Write results to ``async_result_dir``.
+        Args:
+            indices (Sequence[int]): Indices to process.
+
+        Returns:
+            List[Any] | None: Flattened results if gathered; otherwise ``None``.
+
+        Raises:
+            RuntimeError: If Dask client creation fails.
+            ValueError: If async sharding parameters are invalid.
+
+        Notes:
+            - When not gathering, each shard writes to
+              ``async_result_dir / f"result-<job_id>.jsonl"`` on its worker.
+        """
+        par_cfg = get_parallel_config()
+        client = make_client(par_cfg)
+        n_workers = par_cfg.get("n_workers", 1)
+        try:
+            chunks = _default_chunk(indices, n_workers)
+            self.async_specs_dir.mkdir(parents=True, exist_ok=True)
+            self.async_result_dir.mkdir(parents=True, exist_ok=True)
+
+            # DictConfig -> dict
+            cfg_dict = OmegaConf.to_container(self.provider.config, resolve=True)
+            main_file = Path(sys.argv[0]).resolve()
+            for p in map(Path, sys.path):
+                try:
+                    rel = main_file.relative_to(p.resolve())
+                    module_name = ".".join(rel.with_suffix("").parts)
+                    break
+                except ValueError:
+                    continue
+
+            provider_cls = f"{module_name}" f".{self.provider.__class__.__name__}"
+            runner_cls = f"{module_name}.{self.__class__.__name__}"
+
+            job_meta = []
+            for rank, chunk in enumerate(chunks):
+                job_id = f"{uuid4().hex[:8]}-r{rank}"
+                result_path = (
+                    self.async_result_dir / f"result-{job_id}.jsonl"
+                ).as_posix()
+
+                spec = AsyncJobSpec(
+                    runner_cls=runner_cls,
+                    provider_cls=provider_cls,
+                    config=cfg_dict,
+                    params=getattr(self.provider, "params", {}) or {},
+                    indices=list(chunk),
+                    world_size=len(chunks),
+                    world_rank=rank,
+                    result_path=str(result_path),
+                    extras={},
+                )
+                spec_path = self.async_specs_dir / f"spec-{job_id}.json"
+                with open(spec_path, "w", encoding="utf-8") as f:
+                    json.dump(convert_paths(asdict(spec)), f, ensure_ascii=False)
```

Comment:
Would it be good to always convert the paths in the config `spec` to absolute? I'm not sure whether relative paths will be a problem when submitting asynchronous jobs.